[transaction] 3.9-SNAPSHOT com.hazelcast.core.HazelcastException: java.io.IOException  Packet not send to

This cluster member shutdown job Fails every time with

```
com.hazelcast.core.HazelcastException: java.io.IOException  Packet not send to
```

https://hazelcast-l337.ci.cloudbees.com/view/shutdown/job/shutdown-map-txn-1phase/75/console


/disk1/jenkins/workspace/shutdown-map-txn-1phase/3.9-SNAPSHOT/2017_06_16-06_09_36

http://54.163.63.218//~jenkins/workspace/shutdown-map-txn-1phase/3.9-SNAPSHOT/2017_06_16-06_09_36


we have this odd generic exception HazelcastException wrapping a IOException "Packet not send"
exception.  the test is a cluster member controled shutdown test, where we would expect to hit

any of 
```
com.hazelcast.spi.exception.TargetDisconnectedException:com.hazelcast.spi.exception.TargetNotMemberException:com.hazelcast.core.HazelcastInstanceNotActiveException:java.lang.IllegalStateException
```



```
HzClient3HZ txn1 hzcmd.txn.map.validate.GetForUpdateSet threadId=7 com.hazelcast.core.HazelcastException: java.io.IOException: Packet not send to [10.0.0.42]:5701
```


```
com.hazelcast.core.HazelcastException: java.io.IOException: Packet not send to [10.0.0.42]:5701
	at com.hazelcast.util.ExceptionUtil.peel(ExceptionUtil.java:94)
	at com.hazelcast.util.ExceptionUtil.peel(ExceptionUtil.java:56)
	at com.hazelcast.util.ExceptionUtil.peel(ExceptionUtil.java:81)
	at com.hazelcast.util.ExceptionUtil.peel(ExceptionUtil.java:56)
	at com.hazelcast.util.ExceptionUtil.peel(ExceptionUtil.java:52)
	at com.hazelcast.util.ExceptionUtil.rethrow(ExceptionUtil.java:105)
	at com.hazelcast.client.proxy.txn.ClientTxnProxy.invoke(ClientTxnProxy.java:71)
	at com.hazelcast.client.proxy.txn.ClientTxnMapProxy.getForUpdate(ClientTxnMapProxy.java:90)
	at hzcmd.txn.map.validate.GetForUpdateSet.timeStep(GetForUpdateSet.java:52)
	at remote.bench.marker.MetricsMarker.flatOut(MetricsMarker.java:53)
	at remote.bench.marker.MetricsMarker.bench(MetricsMarker.java:40)
	at remote.bench.BenchThread.call(BenchThread.java:38)
	at remote.bench.BenchThread.call(BenchThread.java:12)
	at java.util.concurrent.FutureTask.run(FutureTask.java:262)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:473)
	at java.util.concurrent.FutureTask.run(FutureTask.java:262)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.io.IOException: Packet not send to [10.0.0.42]:5701
	at com.hazelcast.client.spi.impl.ClientInvocationServiceImpl.send(ClientInvocationServiceImpl.java:113)
	at com.hazelcast.client.spi.impl.ClientSmartInvocationServiceImpl.invokeOnConnection(ClientSmartInvocationServiceImpl.java:79)
	at com.hazelcast.client.spi.impl.ClientInvocation.invokeOnSelection(ClientInvocation.java:141)
	at com.hazelcast.client.spi.impl.ClientInvocation.retry(ClientInvocation.java:171)
	at com.hazelcast.client.spi.impl.ClientInvocation.run(ClientInvocation.java:159)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:473)
	at java.util.concurrent.FutureTask.run(FutureTask.java:262)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:178)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:292)
	at com.hazelcast.util.executor.LoggingScheduledExecutor$LoggingDelegatingFuture.run(LoggingScheduledExecutor.java:128)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
	at java.lang.Thread.run(Thread.java:748)
	at com.hazelcast.util.executor.HazelcastManagedThread.executeRun(HazelcastManagedThread.java:64)
	at com.hazelcast.util.executor.HazelcastManagedThread.run(HazelcastManagedThread.java:80)
	at ------ submitted from ------.(Unknown Source)
	at com.hazelcast.client.spi.impl.ClientInvocationFuture.resolveAndThrowIfException(ClientInvocationFuture.java:94)
	at com.hazelcast.client.spi.impl.ClientInvocationFuture.resolveAndThrowIfException(ClientInvocationFuture.java:34)
	at com.hazelcast.spi.impl.AbstractInvocationFuture.get(AbstractInvocationFuture.java:155)
	at com.hazelcast.client.proxy.txn.ClientTxnProxy.invoke(ClientTxnProxy.java:69)
	... 12 more
```

--- a/hazelcast-client/src/main/java/com/hazelcast/client/proxy/txn/ClientTransactionUtil.java
+++ b/hazelcast-client/src/main/java/com/hazelcast/client/proxy/txn/ClientTransactionUtil.java
@@ -1,0 +1,56 @@
+/*
+ * Copyright (c) 2008-2017, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.client.proxy.txn;
+
+import com.hazelcast.client.impl.HazelcastClientInstanceImpl;
+import com.hazelcast.client.impl.protocol.ClientMessage;
+import com.hazelcast.client.spi.impl.ClientInvocation;
+import com.hazelcast.nio.Connection;
+import com.hazelcast.transaction.TransactionException;
+import com.hazelcast.util.ExceptionUtil;
+
+import java.util.concurrent.Future;
+
+/**
+ * Contains static method that is used from client transaction classes
+ */
+public final class ClientTransactionUtil {
+
+    private ClientTransactionUtil() {
+
+    }
+
+    /**
+     * Handles the invocation exception for transactions so that users will not see internal exceptions.
+     * More specifically IOException, because in case of a IO problem in ClientInvocation that send to a connection
+     * sends IOException to user. This wraps that exception into a TransactionException.
+     */
+    public static ClientMessage invoke(ClientMessage request, HazelcastClientInstanceImpl client, Connection connection) {
+        try {
+            final ClientInvocation clientInvocation = new ClientInvocation(client, request, connection);
+            final Future<ClientMessage> future = clientInvocation.invoke();
+            return future.get();
+        } catch (Exception e) {
+            RuntimeException runtimeException = ExceptionUtil.peel(e);
+            if (runtimeException instanceof TransactionException) {
+                throw runtimeException;
+            } else {
+                throw new TransactionException(runtimeException);
+            }
+        }
+    }
+}

--- a/hazelcast-client/src/main/java/com/hazelcast/client/proxy/txn/ClientTxnProxy.java
+++ b/hazelcast-client/src/main/java/com/hazelcast/client/proxy/txn/ClientTxnProxy.java
@@ -21,15 +21,10 @@ import com.hazelcast.client.impl.HazelcastClientInstanceImpl;
 import com.hazelcast.client.impl.protocol.ClientMessage;
 import com.hazelcast.client.impl.protocol.codec.ClientDestroyProxyCodec;
 import com.hazelcast.client.spi.ClientTransactionContext;
-import com.hazelcast.client.spi.impl.ClientInvocation;
 import com.hazelcast.nio.serialization.Data;
 import com.hazelcast.partition.strategy.StringPartitioningStrategy;
 import com.hazelcast.spi.serialization.SerializationService;
 import com.hazelcast.transaction.TransactionalObject;
-
-import java.util.concurrent.Future;
-
-import static com.hazelcast.util.ExceptionUtil.rethrow;
 
 abstract class ClientTxnProxy implements TransactionalObject {
 
@@ -61,15 +56,9 @@ abstract class ClientTxnProxy implements TransactionalObject {
     abstract void onDestroy();
 
     final ClientMessage invoke(ClientMessage request) {
-        try {
-            HazelcastClientInstanceImpl client = transactionContext.getClient();
-            ClientConnection connection = transactionContext.getConnection();
-            ClientInvocation invocation = new ClientInvocation(client, request, connection);
-            Future<ClientMessage> future = invocation.invoke();
-            return future.get();
-        } catch (Exception e) {
-            throw rethrow(e);
-        }
+        HazelcastClientInstanceImpl client = transactionContext.getClient();
+        ClientConnection connection = transactionContext.getConnection();
+        return ClientTransactionUtil.invoke(request, client, connection);
     }
 
     String getTransactionId() {

--- a/hazelcast-client/src/main/java/com/hazelcast/client/proxy/txn/TransactionContextProxy.java
+++ b/hazelcast-client/src/main/java/com/hazelcast/client/proxy/txn/TransactionContextProxy.java
@@ -23,7 +23,6 @@ import com.hazelcast.client.spi.impl.ClientTransactionManagerServiceImpl;
 import com.hazelcast.collection.impl.list.ListService;
 import com.hazelcast.collection.impl.queue.QueueService;
 import com.hazelcast.collection.impl.set.SetService;
-import com.hazelcast.core.HazelcastException;
 import com.hazelcast.core.TransactionalList;
 import com.hazelcast.core.TransactionalMap;
 import com.hazelcast.core.TransactionalMultiMap;
@@ -64,7 +63,7 @@ public class TransactionContextProxy implements ClientTransactionContext {
         try {
             connection = transactionManager.connect();
         } catch (Exception e) {
-            throw new HazelcastException("Could not obtain Connection!", e);
+            throw new TransactionException("Could not obtain a connection!", e);
         }
         this.transaction = new TransactionProxy(client, options, connection);
     }

--- a/hazelcast-client/src/main/java/com/hazelcast/client/proxy/txn/TransactionProxy.java
+++ b/hazelcast-client/src/main/java/com/hazelcast/client/proxy/txn/TransactionProxy.java
@@ -22,15 +22,12 @@ import com.hazelcast.client.impl.protocol.ClientMessage;
 import com.hazelcast.client.impl.protocol.codec.TransactionCommitCodec;
 import com.hazelcast.client.impl.protocol.codec.TransactionCreateCodec;
 import com.hazelcast.client.impl.protocol.codec.TransactionRollbackCodec;
-import com.hazelcast.client.spi.impl.ClientInvocation;
 import com.hazelcast.logging.ILogger;
 import com.hazelcast.transaction.TransactionException;
 import com.hazelcast.transaction.TransactionNotActiveException;
 import com.hazelcast.transaction.TransactionOptions;
 import com.hazelcast.util.Clock;
 import com.hazelcast.util.ThreadUtil;
-
-import java.util.concurrent.Future;
 
 import static com.hazelcast.transaction.impl.Transaction.State;
 import static com.hazelcast.transaction.impl.Transaction.State.ACTIVE;
@@ -84,7 +81,7 @@ final class TransactionProxy {
             startTime = Clock.currentTimeMillis();
             ClientMessage request = TransactionCreateCodec.encodeRequest(options.getTimeoutMillis(),
                     options.getDurability(), options.getTransactionType().id(), threadId);
-            ClientMessage response = invoke(request);
+            ClientMessage response = ClientTransactionUtil.invoke(request, client, connection);
             TransactionCreateCodec.ResponseParameters result = TransactionCreateCodec.decodeResponse(response);
             txnId = result.response;
             state = ACTIVE;
@@ -103,7 +100,7 @@ final class TransactionProxy {
             checkThread();
             checkTimeout();
             ClientMessage request = TransactionCommitCodec.encodeRequest(txnId, threadId);
-            invoke(request);
+            ClientTransactionUtil.invoke(request, client, connection);
             state = COMMITTED;
         } catch (Exception e) {
             state = COMMIT_FAILED;
@@ -122,7 +119,7 @@ final class TransactionProxy {
             checkThread();
             try {
                 ClientMessage request = TransactionRollbackCodec.encodeRequest(txnId, threadId);
-                invoke(request);
+                ClientTransactionUtil.invoke(request, client, connection);
             } catch (Exception exception) {
                 logger.warning("Exception while rolling back the transaction", exception);
             }
@@ -144,13 +141,4 @@ final class TransactionProxy {
         }
     }
 
-    private ClientMessage invoke(ClientMessage request) {
-        try {
-            final ClientInvocation clientInvocation = new ClientInvocation(client, request, connection);
-            final Future<ClientMessage> future = clientInvocation.invoke();
-            return future.get();
-        } catch (Exception e) {
-            throw rethrow(e);
-        }
-    }
 }

--- a/hazelcast-client/src/main/java/com/hazelcast/client/proxy/txn/xa/XATransactionProxy.java
+++ b/hazelcast-client/src/main/java/com/hazelcast/client/proxy/txn/xa/XATransactionProxy.java
@@ -23,7 +23,7 @@ import com.hazelcast.client.impl.protocol.codec.XATransactionCommitCodec;
 import com.hazelcast.client.impl.protocol.codec.XATransactionCreateCodec;
 import com.hazelcast.client.impl.protocol.codec.XATransactionPrepareCodec;
 import com.hazelcast.client.impl.protocol.codec.XATransactionRollbackCodec;
-import com.hazelcast.client.spi.impl.ClientInvocation;
+import com.hazelcast.client.proxy.txn.ClientTransactionUtil;
 import com.hazelcast.logging.ILogger;
 import com.hazelcast.transaction.TransactionException;
 import com.hazelcast.transaction.TransactionNotActiveException;
@@ -33,7 +33,6 @@ import com.hazelcast.util.Clock;
 
 import javax.transaction.xa.XAException;
 import javax.transaction.xa.Xid;
-import java.util.concurrent.Future;
 import java.util.concurrent.TimeUnit;
 
 import static com.hazelcast.transaction.impl.Transaction.State.ACTIVE;
@@ -75,7 +74,7 @@ public class XATransactionProxy {
         try {
             startTime = Clock.currentTimeMillis();
             ClientMessage request = XATransactionCreateCodec.encodeRequest(xid, timeout);
-            ClientMessage response = invoke(request);
+            ClientMessage response = ClientTransactionUtil.invoke(request, client, connection);
             txnId = XATransactionCreateCodec.decodeResponse(response).response;
             state = ACTIVE;
         } catch (Exception e) {
@@ -90,7 +89,7 @@ public class XATransactionProxy {
                 throw new TransactionNotActiveException("Transaction is not active");
             }
             ClientMessage request = XATransactionPrepareCodec.encodeRequest(txnId);
-            invoke(request);
+            ClientTransactionUtil.invoke(request, client, connection);
             state = PREPARED;
         } catch (Exception e) {
             state = ROLLING_BACK;
@@ -109,7 +108,7 @@ public class XATransactionProxy {
             }
             state = COMMITTING;
             ClientMessage request = XATransactionCommitCodec.encodeRequest(txnId, onePhase);
-            invoke(request);
+            ClientTransactionUtil.invoke(request, client, connection);
             state = COMMITTED;
         } catch (Exception e) {
             state = COMMIT_FAILED;
@@ -121,7 +120,7 @@ public class XATransactionProxy {
         state = ROLLING_BACK;
         try {
             ClientMessage request = XATransactionRollbackCodec.encodeRequest(txnId);
-            invoke(request);
+            ClientTransactionUtil.invoke(request, client, connection);
         } catch (Exception exception) {
             logger.warning("Exception while rolling back the transaction", exception);
         }
@@ -143,13 +142,4 @@ public class XATransactionProxy {
         }
     }
 
-    private ClientMessage invoke(ClientMessage request) {
-        try {
-            final ClientInvocation clientInvocation = new ClientInvocation(client, request, connection);
-            final Future<ClientMessage> future = clientInvocation.invoke();
-            return future.get();
-        } catch (Exception e) {
-            throw rethrow(e);
-        }
-    }
 }

--- a/hazelcast-client/src/test/java/com/hazelcast/client/txn/ClientTxnMapTest.java
+++ b/hazelcast-client/src/test/java/com/hazelcast/client/txn/ClientTxnMapTest.java
@@ -254,7 +254,7 @@ public class ClientTxnMapTest {
                     tryPutResult.set(result);
 
                     afterTryPutResult.countDown();
-                } catch (Exception e) {
+                } catch (InterruptedException e) {
                 }
             }
         };
@@ -267,7 +267,7 @@ public class ClientTxnMapTest {
                     txMap.getForUpdate(key);
                     getKeyForUpdateLatch.countDown();
                     afterTryPutResult.await(30, TimeUnit.SECONDS);
-                } catch (Exception e) {
+                } catch (InterruptedException e) {
                 }
                 return true;
             }

--- a/hazelcast-client/src/test/java/com/hazelcast/client/txn/ClientTxnMultiMapTest.java
+++ b/hazelcast-client/src/test/java/com/hazelcast/client/txn/ClientTxnMultiMapTest.java
@@ -24,6 +24,7 @@ import com.hazelcast.test.HazelcastParallelClassRunner;
 import com.hazelcast.test.annotation.ParallelTest;
 import com.hazelcast.test.annotation.QuickTest;
 import com.hazelcast.transaction.TransactionContext;
+import com.hazelcast.transaction.TransactionException;
 import org.junit.After;
 import org.junit.Before;
 import org.junit.Test;
@@ -131,7 +132,7 @@ public class ClientTxnMultiMapTest {
                         context.commitTransaction();
 
                         assertEquals(3, multiMap.get(key).size());
-                    } catch (Exception e) {
+                    } catch (TransactionException e) {
                         error.compareAndSet(null, e);
                     } finally {
                         latch.countDown();

--- a/hazelcast-client/src/test/java/com/hazelcast/client/txn/ClientTxnTest.java
+++ b/hazelcast-client/src/test/java/com/hazelcast/client/txn/ClientTxnTest.java
@@ -30,6 +30,7 @@ import com.hazelcast.test.HazelcastTestSupport;
 import com.hazelcast.test.annotation.ParallelTest;
 import com.hazelcast.test.annotation.QuickTest;
 import com.hazelcast.transaction.TransactionContext;
+import com.hazelcast.transaction.TransactionException;
 import com.hazelcast.transaction.TransactionOptions;
 import org.junit.After;
 import org.junit.Before;
@@ -102,7 +103,7 @@ public class ClientTxnTest extends HazelcastTestSupport {
 
             context.commitTransaction();
             fail("commit should throw exception!!!");
-        } catch (Exception e) {
+        } catch (TransactionException e) {
             context.rollbackTransaction();
             txnRollbackLatch.countDown();
         }
@@ -138,7 +139,7 @@ public class ClientTxnTest extends HazelcastTestSupport {
         try {
             context.commitTransaction();
             fail("commit should throw exception !");
-        } catch (Exception e) {
+        } catch (TransactionException e) {
             context.rollbackTransaction();
             txnRollbackLatch.countDown();
         }
@@ -161,11 +162,15 @@ public class ClientTxnTest extends HazelcastTestSupport {
         TransactionContext context = client.newTransactionContext(options);
         context.beginTransaction();
         try {
-            context.getQueue(name).take();
+            try {
+                context.getQueue(name).take();
+            } catch (InterruptedException e) {
+                fail();
+            }
             sleepAtLeastSeconds(5);
             context.commitTransaction();
             fail();
-        } catch (Exception e) {
+        } catch (TransactionException e) {
             context.rollbackTransaction();
         }
         assertEquals("Queue size should be 1", 1, queue.size());

--- a/hazelcast/src/main/java/com/hazelcast/core/TransactionalMap.java
+++ b/hazelcast/src/main/java/com/hazelcast/core/TransactionalMap.java
@@ -57,7 +57,7 @@ import java.util.concurrent.TimeUnit;
  *
  *      transaction.delistResource(xaResource, XAResource.TMSUCCESS);
  *      tm.commit();
- * } catch (Throwable t) {
+ * } catch (TransactionException t) {
  *      t.printStackTrace();
  *      transaction.delistResource(xaResource, XAResource.TMFAIL);
  *      tm.rollback();

--- a/hazelcast/src/main/java/com/hazelcast/transaction/TransactionContext.java
+++ b/hazelcast/src/main/java/com/hazelcast/transaction/TransactionContext.java
@@ -21,6 +21,8 @@ import javax.transaction.xa.XAResource;
 /**
  * Provides a context to perform transactional operations: beginning/committing transactions, but also retrieving
  * transactional data-structures like the {@link com.hazelcast.core.TransactionalMap}.
+ * Any method accessed through TransactionContext interface can throw TransactionException if transaction is
+ * no longer valid and rolled back.
  */
 public interface TransactionContext extends TransactionalTaskContext {
 

